### ISSUE TITLE
systemd-networkd won't renew DHCPv4 address after NDisc error

### ISSUE THREAD

**Issue by Malvineous** [systemd/systemd#5625]
### Submission type

  - [X] Bug report
  - [ ] Request for enhancement (RFE)

### systemd version the issue has been seen with

systemd 232

### Used distribution

Arch Linux ARM

### In case of bug report: Expected behaviour you didn't see

When DHCPv4 lease expires, the address should be renewed from the DHCP server.

### In case of bug report: Unexpected behaviour you saw

When DHCPv4 lease expired, the address was not renewed and the machine lost IPv4 connectivity.  IPv6 connectivity remained.

### In case of bug report: Steps to reproduce the problem

I am not sure how to reproduce the problem as I do not know what is causing the error message.

I see this in my logs when starting `systemd-networkd`:

```
systemd-networkd[245]: Enumeration completed
systemd-networkd[245]: eth0: IPv6 enabled for interface: Success
systemd-networkd[245]: eth0: Gained carrier
systemd-networkd[245]: eth0: Lost carrier
systemd-networkd[245]: eth0: Gained IPv6LL
systemd-networkd[245]: eth0: Gained carrier
systemd-networkd[245]: eth0: DHCPv4 address 192.168.0.25/24 via 192.168.0.10
systemd-networkd[245]: eth0: Could not set NDisc route or address: Invalid argument
systemd-networkd[245]: eth0: Failed
```
Then 24 hours later (the length of the IPv4 lease) the IPv4 address is removed again with nothing written to the logs.  This has happened on three different machines.

`systemctl status systemd-networkd` shows it as active and running.

Restarting `systemd-networkd` restores the IPv4 address, and displays the following log messages:
```
systemd-networkd[750]: eth0: Gained IPv6LL
systemd-networkd[750]: Enumeration completed
systemd-networkd[750]: eth0: DHCPv4 address 192.168.0.25/24 via 192.168.0.10
systemd-networkd[750]: eth0: Could not set NDisc route or address: Invalid argument
systemd-networkd[750]: eth0: Failed
```
I presume in another 24 hours the IPv4 address will drop out again.

I am not sure what it causing the `NDisc` error message.  I have IPv6 enabled on this network so it may be related to that, however any IPv6 errors should not prevent IPv4 addresses from being renewed.

Please advise if there are further troubleshooting steps I can take to help narrow down the problem.

Here is the content of my `/etc/systemd/network/default.network` file, the only one I have in that directory:
```
[Match]
Name=eth*

[Network]
DHCP=v4

Address=fdaa::ba27:ebff:fe48:1c6b

[DHCPv4]
UseHostname=false
```

**Comment by Malvineous** [comment]
Just confirming that 24 hours later the machines again lost their IP addresses, so `systemd-networkd` is definitely not renewing the address from the DHCP server.

**Comment by hendrikw01** [comment]
It probably is the ipv6 Router Advertisement cause your networkd-[iface] service to fail and therefore the dhcpv4 client is killed too.

In my case, it is caused by my ISP. He cuts the connection every 24h.

You could run systemd with debug messages and look for `Received Router Advertisement`. 

If this is the case, you can try the following workaround to switch ipv6 off:
```
 [Network]
 DHCP=ipv4
 IPv6AcceptRA=no
 LinkLocalAddressing=no
```

Or try my pull request: #5636

**Comment by Malvineous** [comment]
Thanks for the tips!  I do use link local addressing though (it's how I can renew the IPv4 address when it goes away!) and I use IPv6 RAs to get global IPs assigned to the devices, so if I switch off the RAs I'll basically end up as IPv4 only, when I do want to have IPv6 enabled as well.

Is there a way to get debug messages for just `systemd-networkd`?  Looking at how to run with debug messages leads me to a [freedesktop.org page](https://freedesktop.org/wiki/Software/systemd/Debugging/), which suggests changing your kernel command line and rebooting which is a bit of a pain and I don't think the boot messages will be relevant.  Can I enable debugging for just `systemd-networkd`?

I'll also keep an eye out to see if the pull request gets merged so I can test it.  The system in question is ARM and it's not so easy for me to locally compile or cross compile just at the moment.

For the time being my workaround is to run this in a shell on the device:

    while true; do systemctl restart systemd-networkd; sleep 23h; done

So far this has fixed the problem.

**Comment by Malvineous** [comment]
Sorry realised I should be searching for `systemd-networkd debug` not just `systemd debug`.  Here's (hopefully) the debug log from restarting `systemd-networkd`.

Is there a way to find out what's happening in order to produce the message `eth0: Could not set NDisc route or address: Invalid argument` towards the end?

Of course I guess this bug is about not failing when that happens.

```
systemd[1]: Starting Network Service...
systemd-networkd[319]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 error=n/a
systemd-networkd[319]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.24 object=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 error=n/a
systemd-networkd[319]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=2 reply_cookie=0 error=n/a
systemd-networkd[319]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.24 object=n/a interface=n/a member=n/a cookie=3 reply_cookie=2 error=n/a
systemd-networkd[319]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RequestName cookie=3 reply_cookie=0 error=n/a
systemd-networkd[319]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.24 object=n/a interface=n/a member=n/a cookie=5 reply_cookie=3 error=n/a
systemd-networkd[319]: Failed to open configuration file '/etc/systemd/networkd.conf': No such file or directory
systemd-networkd[319]: timestamp of '/etc/systemd/network' changed
systemd-networkd[319]: eth0: Flags change: +UP +LOWER_UP +RUNNING +MULTICAST +BROADCAST
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=4 reply_cookie=0 error=n/a
systemd-networkd[319]: eth0: Link 2 added
systemd-networkd[319]: eth0: udev initialized link
systemd-networkd[319]: eth0: Saved original MTU: 1500
systemd-networkd[319]: lo: Flags change: +LOOPBACK +UP +LOWER_UP +RUNNING
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_31 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=5 reply_cookie=0 error=n/a
systemd-networkd[319]: lo: Link 1 added
systemd-networkd[319]: lo: udev initialized link
systemd-networkd[319]: lo: Saved original MTU: 0
systemd-networkd[319]: eth0: Adding address: fe80::ba27:ebff:fe48:1c6b/64 (valid forever)
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=6 reply_cookie=0 error=n/a
systemd-networkd[319]: eth0: Gained IPv6LL
systemd-networkd[319]: eth0: Adding address: fdaa::ba27:ebff:fe48:1c6b/0 (valid forever)
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=7 reply_cookie=0 error=n/a
systemd-networkd[319]: eth0: Adding address: 2001:4479:cbce:f309:ba27:ebff:fe48:1c6b/64 (valid forever)
systemd-networkd[319]: eth0: Adding address: fdaa:f00b::ba27:ebff:fe48:1c6b/64 (valid forever)
systemd-networkd[319]: lo: Adding address: ::1/128 (valid forever)
systemd-networkd[319]: eth0: Adding address: 192.168.0.25/24 (valid for 23h 58min 40s)
systemd-networkd[319]: lo: Adding address: 127.0.0.1/8 (valid forever)
systemd-networkd[319]: Enumeration completed
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=8 reply_cookie=0 error=n/a
systemd-networkd[319]: eth0: Link state is up-to-date
systemd-networkd[319]: No virtualization found in DMI
systemd-networkd[319]: No virtualization found in CPUID
systemd-networkd[319]: Virtualization XEN not found, /proc/xen/capabilities does not exist
systemd-networkd[319]: No virtualization found in /proc/device-tree/*
systemd-networkd[319]: No virtualization found in /proc/cpuinfo.
systemd-networkd[319]: This platform does not support /proc/sysinfo
systemd-networkd[319]: Found VM virtualization none
systemd-networkd[319]: eth0: found matching network '/etc/systemd/network/default.network'
systemd-networkd[319]: LLDP: Started LLDP client
systemd-networkd[319]: eth0: Started LLDP.
systemd-networkd[319]: eth0: Acquiring DHCPv4 lease
systemd-networkd[319]: DHCP CLIENT (0xc4aaabd6): STARTED on ifindex 2
systemd-networkd[319]: eth0: Discovering IPv6 routers
systemd-networkd[319]: NDISC: Started IPv6 Router Solicitation client
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=9 reply_cookie=0 error=n/a
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=10 reply_cookie=0 error=n/a
systemd-networkd[319]: eth0: Setting addresses
systemd-networkd[319]: NDISC: Sent Router Solicitation
systemd-networkd[319]: DHCP CLIENT (0xc4aaabd6): DISCOVER
systemd-networkd[319]: DHCP CLIENT (0xc4aaabd6): OFFER
systemd-networkd[319]: DHCP CLIENT (0xc4aaabd6): REQUEST (requesting)
systemd-networkd[319]: DHCP CLIENT (0xc4aaabd6): received message was not an ACK, ignoring
systemd-networkd[319]: DHCP CLIENT (0xc4aaabd6): ACK
systemd-networkd[319]: DHCP CLIENT (0xc4aaabd6): lease expires in 23h 59min 58s
systemd-networkd[319]: DHCP CLIENT (0xc4aaabd6): T2 expires in 20h 59min 57s
systemd-networkd[319]: DHCP CLIENT (0xc4aaabd6): T1 expires in 11h 59min 59s
systemd-networkd[319]: eth0: DHCPv4 address 192.168.0.25/24 via 192.168.0.10
systemd-networkd[319]: NDISC: Received Router Advertisement: flags MANAGED preference medium lifetime 65535 sec
systemd-networkd[319]: NDISC: Invoking callback for 'r'.
systemd-networkd[319]: DHCPv6 CLIENT: Started in Managed mode
systemd-networkd[319]: eth0: Acquiring DHCPv6 lease on NDisc request
systemd-networkd[319]: DHCPv6 CLIENT: Sent SOLICIT
systemd-networkd[319]: DHCPv6 CLIENT: Next retransmission in 1s
systemd-networkd[319]: lo: Link state is up-to-date
systemd-networkd[319]: lo: Unmanaged
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_31 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=11 reply_cookie=0 error=n/a
systemd-networkd[319]: DHCPv6 CLIENT: Recv ADVERTISE
systemd-networkd[319]: eth0: Removing address: fdaa::ba27:ebff:fe48:1c6b/0 (valid forever)
systemd-networkd[319]: eth0: Removing address: fdaa:f00b::ba27:ebff:fe48:1c6b/64 (valid forever)
systemd-networkd[319]: eth0: Removing address: 192.168.0.25/24 (valid for 23h 58min 40s)
systemd-networkd[319]: eth0: Removing address: 2001:4479:cbce:f309:ba27:ebff:fe48:1c6b/64 (valid forever)
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=12 reply_cookie=0 error=n/a
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=13 reply_cookie=0 error=n/a
systemd-networkd[319]: eth0: Addresses set
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=14 reply_cookie=0 error=n/a
systemd-networkd[319]: eth0: Adding address: 192.168.0.25/24 (valid for 1d)
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=15 reply_cookie=0 error=n/a
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=16 reply_cookie=0 error=n/a
systemd-networkd[319]: eth0: Could not set NDisc route or address: Invalid argument
systemd-networkd[319]: eth0: Failed
systemd-networkd[319]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=17 reply_cookie=0 error=n/a
systemd-networkd[319]: DHCP CLIENT (0xc4aaabd6): STOPPED
systemd-networkd[319]: NDISC: Stopping IPv6 Router Solicitation client
systemd-networkd[319]: Got message type=signal sender=org.freedesktop.DBus destination=:1.24 object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired cookie=2 reply_cookie=0 error=n/a
systemd-networkd[319]: Got message type=signal sender=org.freedesktop.DBus destination=:1.24 object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired cookie=4 reply_cookie=0 error=n/a
systemd[1]: Started Network Service.
systemd-networkd[319]: eth0: Adding address: 2001:4479:cbce:f309:ba27:ebff:fe48:1c6b/64 (valid forever)
systemd-networkd[319]: eth0: Adding address: fdaa::ba27:ebff:fe48:1c6b/0 (valid forever)
systemd-networkd[319]: eth0: Adding address: fdaa:f00b::ba27:ebff:fe48:1c6b/64 (valid forever)
```

**Comment by hendrikw01** [comment]
This error `Invalid argument` is thrown from the Linux kernel but unfortunately, there is no easy way to tell which argument is invalid. This message `Could not set NDisc route or address:` is only used for ipv6 RA logic so you have to log All RA messages with radvdump( in radvd package ), especially if your prefix is going to change.

**Comment by hendrikw01** [comment]
If you use Arch Linux ARM with armv7h then you can try my version: [systemd-232-99-armv7h.pkg.tar.xz](https://drive.google.com/file/d/0B6dlB0Y8FHcJZUtqeFZROFlJOU0/view?usp=sharing).

**Comment by Malvineous** [comment]
Thanks for the link - `armv6h` unfortunately though (Raspberry Pi).  Is there any way to figure out what command is being run (assuming it's a command being executed) when the `Invalid argument` response is generated?  If I can fiddle with parameters maybe I can work out what the problem is.  Although I guess the real issue is that this IPv6 error shouldn't break IPv4.

This is what `radvdump` says when restarting `systemd-networkd`:
```
#
# radvd configuration generated by radvdump 2.16
# based on Router Advertisement from fe80::6666:b3ff:feb0:9a60
# received by interface eth0
#

interface eth0
{
	AdvSendAdvert on;
	# Note: {Min,Max}RtrAdvInterval cannot be obtained with radvdump
	AdvManagedFlag on;
	AdvOtherConfigFlag on;
	AdvReachableTime 0;
	AdvRetransTimer 0;
	AdvCurHopLimit 64;
	AdvDefaultLifetime 65535;
	AdvHomeAgentFlag off;
	AdvDefaultPreference medium;
	AdvSourceLLAddress on;
	AdvLinkMTU 1280;
	AdvIntervalOpt on;

	prefix 2001:4479:cbce:f309::/64
	{
		AdvValidLifetime infinity; # (0xffffffff)
		AdvPreferredLifetime infinity; # (0xffffffff)
		AdvOnLink on;
		AdvAutonomous on;
		AdvRouterAddr off;
	}; # End of prefix definition


	prefix fdaa:f00b::/64
	{
		AdvValidLifetime infinity; # (0xffffffff)
		AdvPreferredLifetime infinity; # (0xffffffff)
		AdvOnLink on;
		AdvAutonomous on;
		AdvRouterAddr off;
	}; # End of prefix definition


	prefix fdaa::/64
	{
		AdvValidLifetime infinity; # (0xffffffff)
		AdvPreferredLifetime infinity; # (0xffffffff)
		AdvOnLink on;
		AdvAutonomous on;
		AdvRouterAddr off;
	}; # End of prefix definition


	RDNSS fdaa::baae:edff:fe72:7d5f fdaa::7e05:7ff:fe0f:dcaa
	{
		AdvRDNSSLifetime 18000;
	}; # End of RDNSS definition


	DNSSL my.domain.here
	{
		AdvDNSSLLifetime 18000;
	}; # End of DNSSL definition

}; # End of interface definition
```

**Comment by mikken** [comment]
I have s similar case.
eth0 is configured by systemd-networkd and uses DHCPv4 for IPv4 and SLAAC for IPv6.
Normally nothing's wrong.
Once I start a container with systemd-nspawn, the container is attached to br0 (eth0 is not in br0), this is what I see:
```
systemd-networkd[287]: vb-elmes1: Gained carrier
systemd-networkd[287]: br0: Gained carrier
systemd-networkd[287]: eth0: Could not set NDisc route or address: Invalid argument
systemd-networkd[287]: eth0: Failed
```

And then DHCPv4 client does not renew address for eth0 after a few hours.
Restarting systemd-networkd triggers `NDisc` error again.

**Comment by pmhahn** [comment]
I'm using Debian-Stretch with systemd-232-25+deb9u1.
I switched from using "ifupdown" to systemd-network and now have the problem, that my host vanished from the network sometimes: the host is no longer reachable vie IPv6, but sometimes(?) IPv4 still works.
I was using "DHCP=yes" and found the following messages today:
> journalctl -b 0 SYSLOG_IDENTIFIER=systemd-networkd
...
> Sep 17 07:34:10 barrel systemd-networkd[251]: Enumeration completed
> Sep 17 07:34:11 barrel systemd-networkd[251]: eth0: Renamed to enp3s0
> Sep 17 07:34:11 barrel systemd-networkd[251]: enp3s0: IPv6 enabled for interface: Success
> Sep 17 07:34:15 barrel systemd-networkd[251]: enp3s0: Gained carrier
> Sep 17 07:34:16 barrel systemd-networkd[251]: enp3s0: Gained IPv6LL
> Sep 17 07:34:16 barrel systemd-networkd[251]: enp3s0: DHCPv6 address 2003:e2:73f4:6200:2a92:4aff:fe34:4885/128 timeout preferred 3600 valid 7200
> Sep 17 07:34:18 barrel systemd-networkd[251]: enp3s0: DHCPv4 address 192.168.178.33/24 via 192.168.178.1
> Sep 17 07:34:18 barrel systemd-networkd[251]: enp3s0: Configured
> Sep 17 08:05:33 barrel systemd-networkd[251]: enp3s0: DHCPv6 address 2003:e2:73f4:6200:2a92:4aff:fe34:4885/128 timeout preferred 3600 valid 7200
> Sep 17 08:37:35 barrel systemd-networkd[251]: enp3s0: DHCPv6 address 2003:e2:73f4:6200:2a92:4aff:fe34:4885/128 timeout preferred 3600 valid 7200
> Sep 17 08:50:34 barrel systemd-networkd[251]: enp3s0: Could not set NDisc route or address: Invalid argument
> Sep 17 08:50:34 barrel systemd-networkd[251]: enp3s0: Failed

At 10:50 the host disconnected and IPv6 was no longer working, which matches the 2h timeout after 8:50.

I now switched to
> DHCP=ipv4
> IPv6AcceptRA=yes
> Environment=SYSTEMD_LOG_LEVEL=debug

and will see, if that makes a difference.

**Comment by ssahani** [comment]
in `ndisc_netlink_handler` if route or address fails we stop the clients. 
`link_enter_failed->link_stop_clients that is dhcp, ndisc` etc.

 In this case I think we should not stop rather clients should be running.

**Comment by ssahani** [comment]
Fix is waiting on #6918 .
